File: leetcode/leecode174/solution.ts
/**
 * 考点：数组、动态规划、矩阵
 * 题目：Dungeon Game（地下城游戏）
 * 题目描述：骑士从左上角出发救右下角的公主，每个格子有生命增减值。求骑士能成功救公主的最小初始生命值。
 * 示例：dungeon=[[-2,-3,3],[-5,-10,1],[10,30,-5]] 输出 7
 * 思路：反向 DP，dp[i][j]=max(1, min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j])，从右下向左上递推。
 * 时间复杂度：O(m × n)
 * 空间复杂度：O(n) 优化为一维数组
 */
function calculateMinimumHP(dungeon: number[][]): number {
    const m = dungeon.length;
    const n = dungeon[0].length;
    const dp: number[] = new Array(n + 1).fill(Infinity);
    dp[n - 1] = 1;

    for (let i = m - 1; i >= 0; i--) {
        for (let j = n - 1; j >= 0; j--) {
            dp[j] = Math.max(1, Math.min(dp[j], dp[j + 1]) - dungeon[i][j]);
        }
    }

    return dp[0];
}
export { calculateMinimumHP };

File: leetcode/leecode174/solution.py
"""
考点：数组、动态规划、矩阵
题目：Dungeon Game（地下城游戏）
思路：反向 DP，从右下向左上递推。dp[i][j] = max(1, min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j])。
      一维数组优化空间。
时间复杂度：O(m × n)
空间复杂度：O(n)
"""


def calculateMinimumHP(dungeon: list[list[int]]) -> int:
    m = len(dungeon)
    n = len(dungeon[0])
    # dp[j] 表示从当前位置到终点所需的最小生命值
    # 初始化为正无穷，dp[n] 作为哨兵，dp[n-1] 设为 1
    dp: list[float] = [float("inf")] * (n + 1)
    dp[n - 1] = 1  # 到达公主时至少需要 1 点生命

    # 从下到上、从右到左反向递推
    for i in range(m - 1, -1, -1):
        for j in range(n - 1, -1, -1):
            # 选择右边或下边所需生命值较小的路径
            # 减去当前格的值（正数为加血，负数为扣血）
            need = min(dp[j], dp[j + 1]) - dungeon[i][j]
            # 生命值至少为 1
            dp[j] = max(1, need)

    return int(dp[0])


if __name__ == "__main__":
    # 示例: dungeon=[[-2,-3,3],[-5,-10,1],[10,30,-5]] → 7
    dungeon = [[-2, -3, 3], [-5, -10, 1], [10, 30, -5]]
    assert calculateMinimumHP(dungeon) == 7
    print("全部测试通过")
